extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -0,0 +1,2040 @@
+                                           new StagePartition(
+                                               queryKernel.getStageDefinition(finalStageId).getId(),
+                                               readablePartition.getPartitionNumber()
+                                           )
+                                       )
+                                   );
+                                 }
+                                 catch (IOException e) {
+                                   throw new RuntimeException(e);
+                                 }
+                               }
+                           ).collect(Collectors.toList())
+          ).flatMap(
+              frame -> {
+                final Cursor cursor = FrameProcessors.makeCursor(
+                    frame,
+                    queryKernel.getStageDefinition(finalStageId).getFrameReader()
+                );
+
+                final ColumnSelectorFactory columnSelectorFactory = cursor.getColumnSelectorFactory();
+                final ColumnMappings columnMappings = task.getQuerySpec().getColumnMappings();
+                @SuppressWarnings("rawtypes")
+                final List<ColumnValueSelector> selectors =
+                    columnMappings.getMappings()
+                                  .stream()
+                                  .map(
+                                      mapping ->
+                                          columnSelectorFactory.makeColumnValueSelector(
+                                              mapping.getQueryColumn())
+                                  ).collect(Collectors.toList());
+
+                final List<Object[]> retVal = new ArrayList<>();
+                while (!cursor.isDone()) {
+                  final Object[] row = new Object[columnMappings.getMappings().size()];
+                  for (int i = 0; i < row.length; i++) {
+                    row[i] = selectors.get(i).getObject();
+                  }
+                  retVal.add(row);
+                  cursor.advance();
+                }
+
+                return Sequences.simple(retVal);
+              }
+          ).withBaggage(resultReaderExec::shutdownNow)
+      );
+    } else {
+      return null;
+    }
+  }
+
+  private void publishSegmentsIfNeeded(
+      final QueryDefinition queryDef,
+      final ControllerQueryKernel queryKernel
+  ) throws IOException
+  {
+    if (queryKernel.isSuccess() && MSQControllerTask.isIngestion(task.getQuerySpec())) {
+      final StageId finalStageId = queryKernel.getStageId(queryDef.getFinalStageDefinition().getStageNumber());
+
+      //noinspection unchecked
+      @SuppressWarnings("unchecked")
+      final Set<DataSegment> segments = (Set<DataSegment>) queryKernel.getResultObjectForStage(finalStageId);
+      log.info("Query [%s] publishing %d segments.", queryDef.getQueryId(), segments.size());
+      publishAllSegments(segments);
+    }
+  }
+
+  private void cleanUpDurableStorageIfNeeded()
```

Comment:
What if the code where to crash _right here_, before we execute the body of this function. Do we have logic to clean up the resulting orphaned deep storage files elsewhere? If not, should we? The same is true for any in-flight segments just before we publish, etc.

Reply:
Ah, this is part of the reason that durable storage isn't enabled by default. There is the gap you mentioned. I'll add comments so people don't have to consider this for themselves: they can know for sure that the files aren't cleaned up.

> The same is true for any in-flight segments just before we publish, etc.

These, IIRC, are handled by a pending segment cleanup job that exists at the whole-system level (not particular to this extension).